tests/unit/has.ts

```diff
@@ -135,6 +135,50 @@ registerSuite({
 			}, TypeError, 'exists and overwrite not true');
 		},
 
+		'works with thenable'(this: any) {
+			const dfd = this.async();
+
+			const thenable = {
+				then(resolve: (_: number) => void) {
+					setTimeout(() => {
+						resolve(5);
+					}, 10);
+
+					return thenable;
+				}
+			};
+
+			hasAdd('thenable', thenable);
+			assert.isFalse(has('thenable'));
+
+			setTimeout(dfd.callback(() => {
+				assert.equal(has('thenable'), 5);
+			}), 100);
+		},
+
+		'failed thenable removes itself from cache'(this: any) {
+			const dfd = this.async();
+
+			const thenable = {
+				then(_: (_: number) => void, reject: (_: Error) => void) {
+					setTimeout(() => {
+						reject(new Error('test error'));
+					}, 10);
+
+					return thenable;
+				}
+			};
+
+			hasAdd('thenable', thenable);
+			assert.isFalse(has('thenable'));
+
+			setTimeout(dfd.callback(() => {
```

Comment:
So.. if I do anything with promises I'll have to pull in dojo-shim, which isn't currently a dependency. You OK with that?